contracts/MarketMaker.sol

```diff
@@ -54,43 +59,70 @@ contract MarketMaker is Ownable, IERC1155TokenReceiver {
         collateralToken = _collateralToken;
         conditionId = _conditionId;
         fee = _fee;
+
+        require(collateralToken.transferFrom(tx.origin, this, _funding) && collateralToken.approve(pmSystem, _funding));
```

Comment:
Regarding `tx.origin`, see:

* https://github.com/ethereum/solidity/issues/683

PeterBorah says:

> Regardless of the outcome of this pull request, double and triple check that you're actually secure there. Chances are, there are ways for malicious contracts to do unexpected things by impersonating your users or your contracts. Is it ok if an attacker is able to call any of your functions at any time, with any arguments, while appearing to be a legitimate user? If not, tx.origin isn't safe.

> Also consider whether it's ok that your users **can't use multisig wallets, and can't be DAOs or anything like that.**

Anyway, even if there is no security issue in this context, that last bit is probably a show stopper... You can't get around this by doing a two transfers neither (at least not until CREATE2 hits the main chain).

Adding the address of the owner to the argument list of the constructor and passing `msg.sender` in the context of the factory method to this level might work though. Consider this scenario:

1. User sends transaction to collateral token contract, giving address of new AMM an allowance.
2. User then sends transaction to factory contract to create AMM instance at that address. This is probably the point where a vulnerability would be hiding:

    An attacker may send a transaction to factory contract to front-run the creation of the AMM, causing user's transaction to fail, and also creating a different AMM at the very same address the user has provided an allowance. This is a risk, to be sure, but it may not be an actual vulnerability, as unless the attacker can then cause the AMM to transfer collateral from the user, the net effect of this would be that the user's transaction fails. So griefing (at the same cost to both parties) can happen, and there is *potentially* a vulnerability depending on the implementation of the AMM.

Reply:
> Could you clarify what you mean by two transfers here?

Transfer to the factory, which then transfers to the instance.

> Are we looking to keep the funding functionality in the constructor with all these quirks, or revert it back to the funding function?

Dunno...